packages/json-rpc-provider-proxy/src/get-sync-provider.ts

```diff
@@ -0,0 +1,132 @@
+  (onMessage) => {
+    // if it's null it means that the consumer has called `disconnect`
+    // of it's a Promise it means that it's being respolved, otherwise it's resolved
+    let provider: Provider | Promise<Provider> | null
+
+    let bufferedMessages: Array<string> = []
+    const pendingResponses = new Set<RequestId>()
+    const subscriptionManager = getSubscriptionManager()
+
+    const onMessageProxy = (message: string) => {
+      let parsed: any
+      try {
+        parsed = JSON.parse(message)
+      } catch (_) {
+        console.error(`Unable to parse incoming message: ${message}`)
+        return
+      }
+
+      if (parsed.id !== undefined) {
+        pendingResponses.delete(parsed.id)
+        subscriptionManager.onResponse(parsed)
+      } else {
+        subscriptionManager.onNotifiaction(parsed)
+      }
+
+      onMessage(message)
+    }
+
+    const send = (message: string) => {
+      if (!provider) return
+
+      const parsed = JSON.parse(message)
+      subscriptionManager.onSent(parsed)
+      if (parsed.id) pendingResponses.add(parsed.id)
+
+      if (provider instanceof Promise) {
+        bufferedMessages.push(message)
+      } else provider.send(message)
+    }
+
+    const onHalt = (): Promise<Provider> => {
+      bufferedMessages = []
+      const pendingResponsesCopy = [...pendingResponses]
+      pendingResponses.clear()
+
+      // It needs to restart before sending the errored
+      // responses/notifications because the consumer may
+      // react to those by sending new requests
+      const result = start()
+
+      subscriptionManager.onAbort().forEach((msg) => {
+        onMessage(JSON.stringify(msg))
+      })
+
+      pendingResponsesCopy.forEach((id) => {
+        onMessage(
+          JSON.stringify({
+            jsonrpc: "2.0",
+            error: { code: -32603, message: "Internal error" },
+            id,
+          }),
+        )
+      })
+
+      return result
+    }
+
+    const start = (): Promise<Provider> => {
+      const onResolve = (getProvider: ConnectAsyncProvider) => {
+        let halted = false
+        const _onHalt = () => {
+          if (halted) return
+          halted = true
+          onHalt()
+        }
+        const _onMessageProxy = (msg: string) => {
+          if (halted) return
+          onMessageProxy(msg)
+        }
+
+        const result = getProvider(_onMessageProxy, _onHalt)
+        bufferedMessages.forEach((m) => {
+          result.send(m)
+        })
+        bufferedMessages = []
+        return (provider = result)
+      }
+
+      provider = input().then(onResolve, withMacroTask(onHalt))
+      return provider
+    }
+
+    const disconnect = () => {
+      if (!provider) return
+
+      const finishIt = (input: Provider) => {
+        subscriptionManager.onAbort()
+        pendingResponses.clear()
+        provider = null
+        input.disconnect()
+      }
+
+      if (provider instanceof Promise) {
+        provider.then(finishIt)
```

Comment:
What about `provider.finally(finishIt)`?

Reply:
Well, that would create a problem with TS because the `finishIt` function expects an input (the resolved value), which at the type level it's not guaranteed because `finally` also gets evaluated if the Promise rejects.

Leaving aside the issue with the typings, the code that you are proposing would behave exactly the same because this promise it's guaranteed to [never reject](https://github.com/paritytech/polkadot-api/pull/131/files#diff-ebd8b5c37b2e43e7172b54fe6bc93b2dbc269eca8a616679b6242830ec2c1ec6R100).

However, thanks to this comment I realized that I missed a corner case in here: the Promise should stop re-trying to recover from its rejection once the user has invoked `disconnect`... This is why more often than not I prefer working with Observables (b/c they are cancellable) rather than working with Promises. Anyways, I just [added a new commit](https://github.com/paritytech/polkadot-api/pull/131/commits/a6023e660f7d240edd249150dd86cb1b1539a514) which should address this issue.

Thanks for bringing this up @kratico!